### PR TITLE
[runtime env] Support `.gitignore` exclusion in working dir

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

To make it easier to exclude files in the working directory, `.gitignore` is included to decide which file to be excluded.
Including the excluded terms in the list, the ignore file will also be read.

The excluded list now use `.gitignore` syntax.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/runtime_env.py [hunk 1]
```diff
@@ -111,18 +112,25 @@ def _xor_bytes(left: bytes, right: bytes) -> bytes:
...
         handler: Callable,
+        base: Optional[Path] = None,
 ):
-    if path in excludes:
+    if base is None:
+        base = path
+    path_str = str(path.relative_to(base))
+    if path.is_dir():
+        path_str += "/"
+    print(path_str)
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:124]: remove

File: python/ray/_private/runtime_env.py [hunk 2]
```diff
@@ -181,6 +189,20 @@ def _parse_uri(pkg_uri: str) -> Tuple[Protocol, str]:
...
+    ignore_file = path / ".gitignore"
+    excludes = [
+        e if not Path(e).is_absolute() else str(Path(e).relative_to(path))
+        for e in excludes
+    ]
+    if ignore_file.is_file():
+        with ignore_file.open("r") as f:
+            return PathSpec.from_lines("gitwildmatch",
+                                       f.readlines() + excludes)
+    print("excludes: ", excludes)
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:202]: remove

File: python/requirements.txt [hunk 3]
```diff
@@ -24,6 +24,7 @@ py-spy >= 0.2.0
 pyyaml
 redis >= 3.5.0
 requests
+pathspec
```
Comment (edoakes) [inline python/requirements.txt:27]: We should be really careful about adding new dependencies. Is this strictly necessary? If it is, please tag this PR with api-approval-required and we should also discuss offline

File: python/ray/_private/runtime_env.py [hunk 4]
```diff
@@ -111,18 +112,25 @@ def _xor_bytes(left: bytes, right: bytes) -> bytes:
 
 def _dir_travel(
         path: Path,
-        excludes: Set[Path],
+        excludes: PathSpec,
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:115]: This function is starting to get somewhat complicated -- could you add some unit tests for it?

**Review by edoakes — COMMENTED** [review]
Mostly concerned about the additional dependency

**Comment by edoakes** [comment]
Hmm, it does look like the logic to implement this is quite complex:
https://github.com/cpburnz/python-path-specification/blob/master/pathspec/patterns/gitwildmatch.py

Given that the library is very lightweight I think it's probably ok to add it as a dependency
```
(base) ray/serve % pip install -U pathspec
Requirement already satisfied: pathspec in /Users/eoakes/anaconda3/lib/python3.8/site-packages (0.7.0)
Collecting pathspec
  Using cached pathspec-0.8.1-py2.py3-none-any.whl (28 kB)
Installing collected packages: pathspec
  Attempting uninstall: pathspec
    Found existing installation: pathspec 0.7.0
    Uninstalling pathspec-0.7.0:
      Successfully uninstalled pathspec-0.7.0
Successfully installed pathspec-0.8.1
```

File: python/requirements.txt (hunk 3 above)
Comment (iycheng) [reply python/requirements.txt:27]: > Hmm, it does look like the logic to implement this is quite complex:
> https://github.com/cpburnz/python-path-specification/blob/master/pathspec/patterns/gitwildmatch.py
> 
> Given that the library is very lightweight I think it's probably ok to add it as a dependency
> 
> ```
> (base) ray/serve % pip install -U pathspec
> Requirement already satisfied: pathspec in /Users/eoakes/anaconda3/lib/python3.8/site-packages (0.7.0)
> Collecting pathspec
>   Using cached pathspec-0.8.1-py2.py3-none-any.whl (28 kB)
> Installing collected packages: pathspec
>   Attempting uninstall: pathspec
>     Found existing installation: pathspec 0.7.0
>     Uninstalling pathspec-0.7.0:
>       Successfully uninstalled pathspec-0.7.0
> Successfully installed pathspec-0.8.1
> ```

I think I can make the version of the lib fixed to reduce potential risk.

**Comment by iycheng** [comment]
@edoakes  do you know how to fix the doc test failure? I feel I probably need to add that lib somewhere in the doc.

File: python/requirements.txt (hunk 3 above)
Comment (ericl) [reply python/requirements.txt:27]: @richardliaw would a fixed version pathspec be acceptable to add to core deps?

**Comment by pcmoritz** [comment]
Why don't we just vendor that file from the library? Seems like it is basically not changing, so I think vendoring would be good instead of having a new dependency :)

**Comment by iycheng** [comment]
@pcmoritz  I'm OK with either way. Do you mean make them same as `psutil` kind of thing? I saw files in here `python/ray/thirdparty_files`

**Comment by iycheng** [comment]
I notice psutil's installation is actually [here](https://github.com/ray-project/ray/blob/master/python/setup.py#L264) and the reason is not that we don't want to add extra lib. Do you have some suggestion about it? @pcmoritz 

But actually, when checking that file, I do have one question. If we do things like fetch it from the web and install it locally, is it the same as `pip install`? What's the benefit here?

**Comment by pcmoritz** [comment]
@iycheng I would say if it is only a few files like in this case, let's check it into the Ray repository. The licenses are compatible (https://github.com/cpburnz/python-path-specification#license), we just need to add a note to our own LICENSE file.  And then expose `PathSpec` as `ray.util.PathSpec` or similar.

The way this is currently done with psutil is dangerous, because there will be a different behavior of user code on whether they do
```
import ray
import psutil
```

or

```
import psutil
import ray
```

(they might get different versions depending on which one they use). It is better to avoid that kind of problem alltogether.

**Review by edoakes — APPROVED** [review]
LGTM

**Pull request merged**